### PR TITLE
Feature/patro23-192 - create sign up api endpoint

### PR DESCRIPTION
**Pull request by Kepeczsz**
https://tracker.intive.com/jira/browse/PATRO23-192

### CODE REVIEW

File: src/modules/example/api/Controllers/SignUpController.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	///        "password": "admin1",
+	///        "email": "admin1@gmail.com"
+	///     }
+	/// .</remarks>
+	/// <response code="201">Returns the newly created item.</response>
+	/// <response code="400">If the body is not valid.</response>
+	[ProducesResponseType(StatusCodes.Status201Created)]
+	[ProducesResponseType(StatusCodes.Status400BadRequest)]
+	[HttpPost]
+	public async Task<IActionResult> SignUp([FromBody] CreateUser command)
```
Comment (pzajaczkowski) [inline src/modules/example/api/Controllers/SignUpController.cs:49]: I think we should allow anonymous users to access this endpoint. Currently there is global authentication.

File: src/modules/example/application/User/CreatingUser/CreateUser.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+namespace Intive.Patronage2023.Modules.Example.Application.User.CreatingUser;
+
+/// <summary>
+/// Create user command.
+/// </summary>
+/// <param name="Id">Id.</param>
+/// <param name="Username">Username.</param>
+/// <param name="Password">Password.</param>
+/// <param name="Email">Email.</param>
+public record CreateUser(Guid Id, string Username, string Password, string Email) : ICommand;
```
Comment (pzajaczkowski) [inline src/modules/example/application/User/CreatingUser/CreateUser.cs:13]: "The user is required to enter their email address, password, name (first and last) on the registration page"
Based on the story, we don't need username, but we need firstname and lastname fields.

**Review by pzajaczkowski — CHANGES_REQUESTED** [review]

File: src/modules/example/api/Controllers/SignUpController.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
+using FluentValidation;
+using Intive.Patronage2023.Modules.Example.Application.Example;
+using Intive.Patronage2023.Modules.Example.Application.User.CreatingUser;
+using Intive.Patronage2023.Shared.Abstractions.Commands;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Intive.Patronage2023.Modules.Example.Api.Controllers;
```
Comment (mateuszgiedzinski) [inline src/modules/example/api/Controllers/SignUpController.cs:7]: Place all the stuff directly in api. Maybe in future we will have SecurityModule but for now security should be connected to api.

File: src/modules/example/application/User/CreatingUser/CreateUser.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+using Intive.Patronage2023.Shared.Abstractions.Commands;
+
+namespace Intive.Patronage2023.Modules.Example.Application.User.CreatingUser;
```
Comment (mateuszgiedzinski) [inline src/modules/example/application/User/CreatingUser/CreateUser.cs:3]: Move this to Api project.

File: src/modules/example/api/ExampleModule.cs [hunk 5]
```diff
@@ -27,7 +28,7 @@ public static IServiceCollection AddExampleModule(this IServiceCollection servic
 		services.AddScoped<IExampleRepository, ExampleRepository>();
 		services.AddScoped<IValidator<CreateExample>, CreateExampleValidator>();
 		services.AddScoped<IValidator<GetExamples>, GetExamplesValidator>();
-
+		services.AddScoped<IValidator<CreateUser>, CreateUserValidator>();
```
Comment (mateuszgiedzinski) [inline src/modules/example/api/ExampleModule.cs:31]: Do not mess example module with user creation.

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

**Comment by pzajaczkowski** [comment]
I think you forgot to delete create user from example module

File: src/api/app/Controllers/SignUpController.cs [hunk 6]
```diff
@@ -40,8 +40,8 @@ public SignUpController(ICommandBus commandBus, IValidator<CreateUser> createUse
 	///        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
 	///        "FirstName": "Jan",
 	///        "LastName": "Kowalski",
-	///        "password": "testPasword123!@",
-	///        "email": "jkowalski@gmail.com"
+	///        "Password": "testPasword123!@",
```
Comment (mateuszgiedzinski) [inline src/api/app/Controllers/SignUpController.cs]: we need to use camel case as it is  standard for json.

File: src/api/app/Keycloak/Token/GetToken.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
+using Intive.Patronage2023.Shared.Abstractions.Commands;
+
+namespace Intive.Patronage2023.Api.Keycloak.Token;
+
+/// <summary>
+/// .
```
Comment (mateuszgiedzinski) [inline src/api/app/Keycloak/Token/GetToken.cs]: dot is not valid summary

File: src/api/app/Keycloak/Token/GetToken.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+/// </summary>
+public record GetToken(Guid ClientId) : ICommand;
+
+/// <summary>
+/// .
+/// </summary>
+public class GetTokenHandler : ICommandHandler<GetToken>
```
Comment (mateuszgiedzinski) [inline src/api/app/Keycloak/Token/GetToken.cs]: Command handler and Get seems weird. Commands won't return any results. Maybe this should be query. But why this is added for sign up ?

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

File: src/api/app/User/UserModule.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+{
+	/// <summary>
+	/// Add module services.
+	/// </summary>
+	/// <param name="services">IServiceCollection.</param>
+	/// <param name="configurationManager">ConfigurationManager.</param>
+	/// <returns>Updated IServiceCollection.</returns>
+	public static IServiceCollection AddUserModule(this IServiceCollection services, ConfigurationManager configurationManager)
+	{
+		services.AddDbContext<ExampleDbContext>(options => options.UseSqlServer(configurationManager.GetConnectionString("AppDb")));
```
Comment (mateuszgiedzinski) [inline src/api/app/User/UserModule.cs:22]: User module don't need any  db context for now so please remove this registration.

File: src/modules/example/application/User/CreatingUser/CreateUser.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+/// <summary>
+/// Create user command.
+/// </summary>
+/// <param name="Id">User identifier.</param>
+/// <param name="FirstName">FirstName which user provides.</param>
```
Comment (mateuszgiedzinski) [inline src/modules/example/application/User/CreatingUser/CreateUser.cs:9]: "First name which user provides".

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

File: src/api/app/Controllers/SignUpController.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	///        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
+	///        "firstName": "Jan",
+	///        "lastName": "Kowalski",
+	///        "password": "testPasword123!@",
+	///        "email": "jkowalski@gmail.com"
+	///     }
+	/// .</remarks>
+	/// <response code="201">Returns the newly created user.</response>
+	/// <response code="400">If the body is not valid.</response>
+	[ProducesResponseType(typeof(string), StatusCodes.Status201Created)]
```
Comment (pzajaczkowski) [inline src/api/app/Controllers/SignUpController.cs:49]: Wrong response type. Maybe we need some kind of UserInfo record?

**Review by pzajaczkowski — CHANGES_REQUESTED** [review]

File: src/api/app/Controllers/SignUpController.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
+using FluentValidation;
+using Intive.Patronage2023.Modules.Example.Api.User.CreatingUser;
+using Intive.Patronage2023.Modules.Example.Application.Example;
+using Intive.Patronage2023.Shared.Abstractions.Commands;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Intive.Patronage2023.Modules.Example.Api.Controllers;
```
Comment (DawidBski) [inline src/api/app/Controllers/SignUpController.cs:8]: Wrong namespace

File: src/api/app/Program.cs [hunk 13]
```diff
@@ -1,5 +1,6 @@
 using Intive.Patronage2023.Api.Configuration;
 using Intive.Patronage2023.Modules.Example.Api;
+using Intive.Patronage2023.Modules.Example.Api.User;
```
Comment (DawidBski) [inline src/api/app/Program.cs:3]: wrong using

File: src/api/app/User/CreatingUser/CreateUser.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
+using Intive.Patronage2023.Shared.Abstractions.Commands;
+
+namespace Intive.Patronage2023.Modules.Example.Api.User.CreatingUser;
```
Comment (DawidBski) [inline src/api/app/User/CreatingUser/CreateUser.cs:3]: namespace

File: src/api/app/User/CreatingUser/CreateUserValidator.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
+using FluentValidation;
+
+namespace Intive.Patronage2023.Modules.Example.Api.User.CreatingUser;
```
Comment (DawidBski) [inline src/api/app/User/CreatingUser/CreateUserValidator.cs:3]: change namespace

File: src/api/app/Controllers/SignUpController.cs (hunk 11 above)
Comment (Kepeczsz) [reply src/api/app/Controllers/SignUpController.cs:49]: I have asked leaders if, there should be another response type,  and for now it should be like this.

Comment (mateuszgiedzinski) [reply src/api/app/Controllers/SignUpController.cs:49]: For 201 of course but for 400 no.

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

**Review by pzajaczkowski — APPROVED** [review]

File: src/api/app/Controllers/SignUpController.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	[ProducesResponseType(StatusCodes.Status400BadRequest)]
+	[AllowAnonymous]
+	[HttpPost]
+	public async Task<IActionResult> SignUp([FromBody] CreateUser command)
+	{
+		var validationResult = await this.createUserValidator.ValidateAsync(command);
+		if (validationResult.IsValid)
+		{
+			await this.commandBus.Send(command);
+			return this.Created($"user/{command.FirstName}", command.FirstName);
```
Comment (mateuszmuszynski) [inline src/api/app/Controllers/SignUpController.cs:59]: Why you return firstName here? This is an url to resource that doesn't exist. I think we can just return `this.Created()`

**Review by mateuszmuszynski — CHANGES_REQUESTED** [review]

**Review by mateuszmuszynski — APPROVED** [review]

**Pull request merged**